framework/src/main/java/org/checkerframework/framework/StubTypes.java

```diff
@@ -0,0 +1,476 @@
+package org.checkerframework.framework;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.JarURLConnection;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.TypeElement;
+import javax.tools.Diagnostic.Kind;
+import org.checkerframework.framework.qual.StubFiles;
+import org.checkerframework.framework.source.SourceChecker;
+import org.checkerframework.framework.stub.StubParser;
+import org.checkerframework.framework.stub.StubResource;
+import org.checkerframework.framework.stub.StubUtil;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.ElementUtils;
+import org.checkerframework.javacutil.PluginUtil;
+
+/** Holds information about types parsed from stub files. */
+public class StubTypes {
+    /** Types read from stub files (but not those from the annotated JDK jar file). */
+    private final Map<Element, AnnotatedTypeMirror> typesFromStubFiles;
+
+    /**
+     * Declaration annotations read from stub files (but not those from the annotated JDK jar file).
+     * Map keys cannot be Element, because a different Element appears in the stub files than in the
+     * real files. So, map keys are the verbose element name, as returned by
+     * ElementUtils.getVerboseName.
+     */
+    private final Map<String, Set<AnnotationMirror>> declAnnosFromStubFiles;
+
+    /**
+     * Whether or not a stub file is currently being parsed. (If one is being parsed, don't try to
+     * parse another.)
+     */
+    private boolean parsing;
+
+    /** AnnotatedTypeFactory */
+    private final AnnotatedTypeFactory factory;
+
+    /** JDK stub files from the file system. */
```

Comment:
For this field and the next it would be interesting to know the meaning of the keys/values.